matrix_functions.py

```diff
@@ -381,24 +415,63 @@ def _matrix_inverse_root_eigen(
     # compute eigendecomposition and compute minimum eigenvalue
     L, Q = _eigh_eigenvalue_decomposition(
         A_ridge,
-        retry_double_precision=retry_double_precision,
-        eigendecomposition_offload_device=eigendecomposition_offload_device,
+        retry_double_precision=root_inv_config.retry_double_precision,
+        eigendecomposition_offload_device=root_inv_config.eigendecomposition_offload_device,
     )
-    lambda_min = torch.min(L)
 
-    # make eigenvalues > 0 (if necessary)
+    power_L = scale_and_pow_eigenvalues(
+        L,
+        root,
+        epsilon=epsilon,
+        noninvertible_handling_config=root_inv_config.noninvertible_handling_config,
+    )
+
+    # compute inverse preconditioner
+    X = Q * power_L.unsqueeze(0) @ Q.T
+
+    return X, L, Q
 
-    if enhance_stability:
-        L += -torch.minimum(lambda_min - epsilon, torch.as_tensor(0.0))
+
+def scale_and_pow_eigenvalues(
```

Comment:
Maybe something like `stabilize_and_pow_eigenvalues` is more precise since how we stabilize the computation depends on `noninvertible_handling_config` (either truncate for pseudo-inverse or shift by `epsilon`).